src/api.ts

```diff
@@ -0,0 +1,362 @@
+    onWatcherRerun: (listener: BirpcEvents['onWatcherRerun']) => void
+    clearListeners: () => void
+    removeListener: (name: string, listener: any) => void
+  }
+}
+
+function createHandler<T extends (...args: any) => any>() {
+  const handlers: T[] = []
+  return {
+    handlers,
+    register: (listener: any) => handlers.push(listener),
+    trigger: (...data: any) => handlers.forEach(handler => handler(...data)),
+    clear: () => handlers.length = 0,
+    remove: (listener: T) => {
+      const index = handlers.indexOf(listener)
+      if (index !== -1)
+        handlers.splice(index, 1)
+    },
+  }
+}
+
+function createRpcOptions() {
+  const handlers = {
+    onConsoleLog: createHandler<BirpcEvents['onConsoleLog']>(),
+    onTaskUpdate: createHandler<BirpcEvents['onTaskUpdate']>(),
+    onFinished: createHandler<BirpcEvents['onFinished']>(),
+    onCollected: createHandler<BirpcEvents['onCollected']>(),
+    onWatcherRerun: createHandler<BirpcEvents['onWatcherRerun']>(),
+    onWatcherStart: createHandler<BirpcEvents['onWatcherStart']>(),
+  }
+
+  const events: Omit<BirpcEvents, 'onReady' | 'onError'> = {
+    onConsoleLog: handlers.onConsoleLog.trigger,
+    onFinished: handlers.onFinished.trigger,
+    onTaskUpdate: handlers.onTaskUpdate.trigger,
+    onCollected: handlers.onCollected.trigger,
+    onWatcherRerun: handlers.onWatcherRerun.trigger,
+    onWatcherStart: handlers.onWatcherStart.trigger,
+  }
+
+  return {
+    events,
+    handlers: {
+      onConsoleLog: handlers.onConsoleLog.register,
+      onTaskUpdate: handlers.onTaskUpdate.register,
+      onFinished: handlers.onFinished.register,
+      onCollected: handlers.onCollected.register,
+      onWatcherRerun: handlers.onWatcherRerun.register,
+      onWatcherStart: handlers.onWatcherStart.register,
+      removeListener(name: string, listener: any) {
+        handlers[name as 'onCollected']?.remove(listener)
+      },
+      clearListeners() {
+        for (const name in handlers)
+          handlers[name as 'onCollected']?.clear()
+      },
+    },
+  }
+}
+
+function nonNullable<T>(value: T | null | undefined): value is T {
+  return value != null
+}
+
+interface VitestMeta {
+  folder: vscode.WorkspaceFolder
+  vitestNodePath: string
+  version: string
+}
+
+export function resolveVitestFoldersMeta(folders: readonly vscode.WorkspaceFolder[]): VitestMeta[] {
+  return folders.map((folder) => {
+    const vitestPackagePath = resolveVitestPackagePath(folder)
+    if (!vitestPackagePath)
+      return null
+    const pkg = _require(vitestPackagePath)
+    const vitestNodePath = resolveVitestNodePath(vitestPackagePath)
+    return { folder, vitestNodePath, version: pkg.version }
+  }).filter(nonNullable)
+}
+
+function createChildVitestProcess(meta: VitestMeta[]) {
+  const vitest = fork(
+    workerPath,
+    {
+      // TODO: use user's execPath to use the local node version (also expose an option?)
```

Comment:
Not sure how I feel about TODOs. I feel like they are never followed up on. Can we also create an issue for it, and link it in the TODO, so it is tracked?

Unless of course it's because you're not done yet (as it is still a draft PR).

Reply:
This is a draft, all todos should be resolved before this is merged